bluemira/utilities/positioning.py

```diff
@@ -0,0 +1,369 @@
+        ln = vector_lengthnorm_2d(x, z)
+        self.x_ius = InterpolatedUnivariateSpline(ln, x)
+        self.z_ius = InterpolatedUnivariateSpline(ln, z)
+
+    @staticmethod
+    def _f_min(l_value, f_x, f_z, x, z):
+        dx = f_x(l_value) - x
+        dz = f_z(l_value) - z
+        return dx ** 2 + dz ** 2
+
+    def to_xz(self, l_value):
+        """
+        Convert parametric-space 'L' values to physical x-z space.
+        """
+        return float(self.x_ius(l_value)), float(self.z_ius(l_value))
+
+    def to_L(self, x, z):
+        """
+        Convert physical x-z space values to parametric-space 'L' values.
+        """
+        return minimize_scalar(
+            self._f_min,
+            args=(self.x_ius, self.z_ius, x, z),
+            bounds=[0, 1],
+            method="bounded",
+            options={"xatol": 1e-8},
+        ).x
+
+
+class RegionInterpolator(XZGeometryInterpolator):
+    """
+    Sets up an x-z region for a point to move within.
+
+    The region is treated as a flat x-z surface.
+
+    The normalisation occurs by cutting the shape in two axes and
+    normalising over the cut length within the region.
+
+    Currently this is limited to convex polygons.
+
+    Generalisation to all polygons is possible but unimplemented
+    and possibly quite slow when converting from normalised to real coordinates.
+
+    When the point position provided is outside the given region the point will
+    be moved to the closest edge of the region.
+
+    The mapping from outside to the edge of the region is not strictly defined.
+    The only certainty is that the point will be moved into the region.
+
+    Parameters
+    ----------
+    geometry: BluemiraWire
+        Region to interpolate within
+    """
+
+    def __init__(self, geometry):
+        super().__init__(geometry)
+        self._check_geometry_feasibility(geometry)
+        self.z_min = geometry.bounding_box.z_min
+        self.z_max = geometry.bounding_box.z_max
+
+    def _check_geometry_feasibility(self, geometry):
+        """
+        Checks the provided region is convex.
+
+        This is a current limitation of RegionInterpolator
+        not providing a 'smooth' interpolation surface.
+
+        Parameters
+        ----------
+        geometry: BluemiraWire
+            Region to check
+
+        Raises
+        ------
+        PositionerError
+            When geometry is not a convex
+        """
+        if not self.geometry.is_closed:
+            raise PositionerError("RegionInterpolator can only handle closed wires.")
+
+        xz_coordinates = self._get_xz_coordinates()
+        hull = ConvexHull(xz_coordinates.T)
+        # Yes, the "area" of a 2-D scipy ConvexHull is its perimeter...
+        if not np.allclose(hull.area, geometry.length, atol=EPS):
+            raise PositionerError(
+                f"RegionInterpolator can only handle convex geometries. Perimeter difference between convex hull and geometry: {hull.volume - geometry.area}"
```

Comment:
Dunno how this line got passed the black check, its quite long!

Reply:
I know, right!?